### DISCUSSION TITLE
Strange behavior with treesitter parsers.

### DISCUSSION THREAD

**Discussion by ambaradan** [NvChad/NvChad#2905] [General]
Hi, and as always thank you for your work. Today during some tests of cloning my configuration I noticed an error message right after the installation of the plugins, this error seems to block the next installation of the *treesitters* parsers, the parsers are in fact installed the next time the editor is started. I verified that this error is also present in a *vanilla* installation of NvChad starting from the *starter* repository and again the same situation occurred. The error is as follows:

```lua
Errore/i eseguendo /home/ambaradan/.config/nvtest/init.lua:                                                                                   
Failed to run `config` for nvim-treesitter                                                                                                              
cannot open /home/ambaradan/.local/share/nvtest/nvchad/base46/syntax: File o directory non esistente                                                    
# stacktrace:                                                                                                                                           
  - /NvChad/lua/nvchad/plugins/init.lua:27 _in_ **config**                                                                                              
  - .config/nvtest/init.lua:17
```

The log is localized in Italian but I think you can understand it.

Installed with:

```bash
git clone https://github.com/NvChad/starter ~/.config/nvtest && NVIM_APPNAME=nvtest nvim
```

**Comment by ambaradan** [comment]
This is the result of `:messages` after installing the plugins:

![install_error](https://github.com/NvChad/NvChad/assets/87865413/4b31bbcc-66c6-4bb9-98fa-30695e5456ed)

If I open the *init.lua* file at the end of *MasonInstallAll* (without restarting the editor) I get this:

![open_init_lua](https://github.com/NvChad/NvChad/assets/87865413/dd4d3fe2-2b8c-4efa-80e1-d85661a9554a)

When the error display ends, the *init.lua* file is opened and correctly highlighted.

**Comment by siduck** [comment]
i think this is due to all plugins being installed asynchronously so the treesitter stuff gets setup before base46 build opt runs, i think 

`:lua require('base46').compile()` should fix it

**Reply by ambaradan** [reply]
Yes I think it is as you say, and that is why some test installations had not shown the error, *treesitter* was loaded after *base46*.

**Reply by siduck** [reply]
are you still facing this error? @ambaradan

**Reply by ambaradan** [reply]
Yes with each installation (clone _my repo_ or _starter_), and I have done quite a few tests.

**Reply by ambaradan** [reply]
The problem is solved after a restart of the editor and a call to a command (e.g. `:Telescope`), launched the command you see the compilation procedure of the parsers start and finish correctly, once installed you have no more problems

**Reply by ambaradan** [reply]
The step of closing and opening the editor alone is not enough to start the installation of parsers.

**Reply by siduck** [reply]
https://github.com/NvChad/NvChad/assets/59060246/fd2c9cf5-1c66-44c9-8a01-915cdcf02acb

i can no longer reproduce this tho! 

Also seems like lazy.nvim doesnt do shallow cloning anymore, which would mean all of the plugin repo gets cloned? before it used to be `--depth 1`, thats why plugins installed faster and base46 somehow built before treesitter could load its config function :joy:

**Reply by ambaradan** [reply]
I confirm I just tried an installation from *starter* and no longer had the problem, *base46* was installed after the first two base plugins!

**Comment by ambaradan** [comment]
My configuration instead that has many more plugins to install highlights the problem.
![my_config](https://github.com/NvChad/NvChad/assets/87865413/954e72d6-1075-4fc6-a785-3f32d27e066f)

**Reply by siduck** [reply]
so it doesnt happen with vanilla nvchad? Have you faced this issue before?

**Reply by ambaradan** [reply]
No with the *starter* repo the treesitter plugin installs regularly, only with my configuration, just tried it.

**Reply by ambaradan** [reply]
I need to investigate

**Comment by ambaradan** [comment]
# Debug base46 error

@siduck I spent the entire evening yesterday debugging the problem I encountered and found two errors, one in a plugin I added and one in NvChad although I think they are related (they are both related to `nvim-treesitter`).

## Debugging Methodology

Below is the procedure used for testing, this is to assess whether the method can introduce any differences:

* First installation using starter repository, installation in `.config/nvtest`.
* Removing the data folder `.local/share/nvtest`
* Configuration changes
* New installation using local folder with `NVIM_APPNAME=nvtest nvim`
* Verification of correct installation
* Removing the data folder `.local/share/nvtest`
* New configuration changes
* New installation
* Check

## Result

By proceeding with this method I was able to isolate the part that introduces the `base46` error, it is the options of *nvim-treesitter* in `plugins/init.lua`:

```lua
-- {
-- 	"nvim-treesitter/nvim-treesitter",
-- 	opts = {
-- 		ensure_installed = {
-- 			"vim", "lua", "vimdoc",
--            "html", "css"
-- 		},
-- 	},
-- },
```

The installations ends correctly once the `lspconfig` and `mason` part of the `plugins/init.lua` file is enabled but ends with the *base46* error if I also enable the `nvim-treesitter` options.

The fact that the second error, this one present in my plugin, was a dependency request of the `nvim-treesitter` plugin makes me think that there is some conflict between the configuration of *nvim-treesitter* in NvChad and that of your own configuration.

I don't think a simple:

```lua
dependencies = { "nvim-treesitter/nvim-treesitter" },
```

in the plugin configuration can create such a problem.

## Remarks

I have my doubts that it is an internal problem with NvChad also because it stopped working properly a couple of days ago and until then I have used it many times without a problem. Maybe it is due to some modification of the basic plugins (`lazy.nvim`) but I don't have the knowledge to go looking for it.

I noticed that the *parsers* are duplicated there are two `ensure_installed` tables, one in NvChad and one in the personal configuration that contain the same parsers and I was wondering if this might be a problem. I haven't tried removing the duplicates and reinstalling (as soon as I have time!).

With the *nvim-treesitter* part commented out and the dependency string removed, my configuration also installs regularly.

**Reply by siduck** [reply]
@ambaradan  dont spend your time on this, I know whats causing this issue. Its cuz base46 is being built after treesitter plugin, and this stuff happens with treesitter only cuz it has TSUpdate in its build, no idea if that runs the config function which has the dofile stuff

this didnt happen before cuz lazy before used shallow clones which are faster to clone than partial  clone i think ( it uses partial rn ) 

So i think wrapping that dofile code with pcall is the only fix or building base46 before bootstrap. so wait for a bit, i'll investigate more

**Reply by ambaradan** [reply]
Great, I'm glad you've already found the problem, I've been having fun anyway and exploring the configuration a little more, which doesn't hurt.
I am curious to see the modifications!
Have a great day

**Reply by siduck** [reply]
@ambaradan  can you try now? `rm -rf .local/share/nvim && nvim`

**Reply by ambaradan** [reply]
I still get the error if I install with *nvim-treesitter* options enabled in `plugins/init.lua`.
Installation done from the folder created with the *starter* repo

**Reply by ambaradan** [reply]
@siduck Wait a moment, this was with the old configuration that still had the old NvChad plugin, let me do some tests more calmly but it seems to me that the problem is solved.

**Reply by ambaradan** [reply]
Later when I get home I will try it out better.

**Comment by ambaradan** [comment]
@siduck I can confirm that the change solves the problem with both a *starter* configuration and a larger one like mine. The failed installation above was done with an outdated `lazy-lock.json` file that still pointed to the previous commit in NvChad. The installation from folder, unlike the installation from *starter* installs plugins to the version referring to the repo commit and not to the most updated version and consequently installed the version without the NvChad plugin fix. Once I updated the `lazy-lock` of my configuration I was able to replicate it without any problem, clearly with the *nvim-treesitter* options enabled.
I think it is safe to say that the problem is solved.
Good job
